[JIT] Store schema in serialized modules and check arguments in function call

goldsborough opened the pull request:
This PR adds argument checking for script method invocation from C++. For this I had to:
1. The schema of a method is currently not serialized in script modules, so we now store the function schema in the `doc_string` field of the ONNX proto. Upon loading of a serialized script module, we parse the schema into the structured C++ form and assign it to the loaded method,
2. Inside `Method::operator()`, we now verify the number and types of arguments.

CC @li-roy 

@zdevito

zdevito reviewed (approved):
Looks good. There are two small fixes.

goldsborough committed:
Store schema in serialized modules and check arguments in function call

Add test for missing argument

Always store schema

Add default arguments to method input and warn against default arguments on export

goldsborough committed:
Support parsing tuples in SchemaParser

zdevito commented on torch/csrc/jit/operator.cpp:
```diff
@@ -62,25 +62,48 @@ struct SchemaParser {
+    }
+    types.push_back(std::move(type));
+  }
+  void parseArgumentType(Argument& arg) {
```
Let's unify these and change it to `parseArgumentType(TypePtr* type, size_t* N)`. Then you can implement `parseType(std::vector<TypePtr>& types) { types.push_back(); parseArgumentType(types.back(), nullptr); }`

apaszke commented on torch/csrc/jit/type.h:
```diff
@@ -351,8 +351,8 @@ struct TORCH_API TupleType : public Type {
   template<typename ... T>
-  static TupleTypePtr create( T&& ... all ) {
-    return TupleTypePtr(new TupleType( std::forward<T>(all)... ));
+  static TupleTypePtr create(std::vector<TypePtr> types) {
```
This function is still a template, but doesn't use any of its parameters.
